cartographer/cloud/internal/local_trajectory_uploader.cc

```diff
@@ -182,8 +200,16 @@ void LocalTrajectoryUploader::TryRecovery() {
                                         entry.second.expected_sensor_ids,
                                         entry.second.trajectory_options);
     if (!status.ok()) {
+      if (get_channel_state() == grpc_connectivity_state::GRPC_CHANNEL_READY) {
+        // TODO: Channel state can be ready even if there's no network.
+        // Could it be that the TCP socket still alive?
+        LOG(WARN) << "Request failed, but uplink channel claims to be ready."
+      }
       LOG(ERROR) << "Failed to create trajectory. Aborting recovery attempt. "
                  << status.error_message();
+      // Restore the previous state for the next recovery attempt.
+      local_trajectory_id_to_trajectory_info_ =
+          local_trajectory_id_to_trajectory_info;
       return;
```

Comment:
This is required, otherwise the next recovery attempt won't have any IDs to register.

Reply:
Good catch!